File: AVLTreeVisualizer/AVLTree.java
import java.util.*;

class AVLNode {

    int data;
    int height;
    AVLNode left, right;

    AVLNode(int data){
        this.data = data;
        this.height = 1;
    }
}

public class AVLTree {

    int height(AVLNode n){
        if(n==null) return 0;
        return n.height;
    }

    int getBF(AVLNode n){
        if(n==null) return 0;
        return height(n.left)-height(n.right);
    }

    AVLNode rightRotate(AVLNode y){

        AVLNode x=y.left;
        AVLNode t2=x.right;

        x.right=y;
        y.left=t2;

        y.height=1+Math.max(height(y.left),height(y.right));
        x.height=1+Math.max(height(x.left),height(x.right));

        return x;
    }

    AVLNode leftRotate(AVLNode x){

        AVLNode y=x.right;
        AVLNode t2=y.left;

        y.left=x;
        x.right=t2;

        x.height=1+Math.max(height(x.left),height(x.right));
        y.height=1+Math.max(height(y.left),height(y.right));

        return y;
    }

    AVLNode insert(AVLNode node,int key){

        if(node==null)
            return new AVLNode(key);

        if(key < node.data)
            node.left = insert(node.left,key);
        else
            node.right = insert(node.right,key); // duplicates right

        node.height = 1 + Math.max(height(node.left),height(node.right));

        int bf = getBF(node);

        if(bf>1 && key < node.left.data)
            return rightRotate(node);

        if(bf<-1 && key >= node.right.data)
            return leftRotate(node);

        if(bf>1 && key >= node.left.data){
            node.left=leftRotate(node.left);
            return rightRotate(node);
        }

        if(bf<-1 && key < node.right.data){
            node.right=rightRotate(node.right);
            return leftRotate(node);
        }

        return node;
    }

    void preorder(AVLNode root){

        if(root!=null){
            System.out.print(root.data+" ");
            preorder(root.left);
            preorder(root.right);
        }
    }

    public static void main(String[] args){

        Scanner sc=new Scanner(System.in);

        AVLTree tree=new AVLTree();
        AVLNode root=null;

        System.out.print("Enter number of nodes: ");
        int n=sc.nextInt();

        for(int i=0;i<n;i++){

            int x=sc.nextInt();
            root=tree.insert(root,x);
        }

        System.out.println("\nPreorder traversal:");
        tree.preorder(root);
    }
}

File: avlTree.js
class Node {
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
        this.height = 1;
    }
}

class AVLTree {
    constructor() { this.root = null; }

    getHeight(node) { return node ? node.height : 0; }
    
    getBalanceFactor(node) { 
        return node ? this.getHeight(node.left) - this.getHeight(node.right) : 0; 
    }

    rightRotate(y) {
        let x = y.left;
        let T2 = x.right;
        x.right = y;
        y.left = T2;
        y.height = Math.max(this.getHeight(y.left), this.getHeight(y.right)) + 1;
        x.height = Math.max(this.getHeight(x.left), this.getHeight(x.right)) + 1;
        return x;
    }

    leftRotate(x) {
        let y = x.right;
        let T2 = y.left;
        y.left = x;
        x.right = T2;
        x.height = Math.max(this.getHeight(x.left), this.getHeight(x.right)) + 1;
        y.height = Math.max(this.getHeight(y.left), this.getHeight(y.right)) + 1;
        return y;
    }

    insert(node, value) {
        if (!node) return new Node(value);

        // ALLOW DUPLICATES: If value is <= node.value, it goes left (or right)
        // Here we put duplicates in the RIGHT subtree
        if (value < node.value) {
            node.left = this.insert(node.left, value);
        } else {
            node.right = this.insert(node.right, value);
        }

        node.height = 1 + Math.max(this.getHeight(node.left), this.getHeight(node.right));
        let balance = this.getBalanceFactor(node);

        // Balancing Logic
        if (balance > 1 && value < node.left.value) return this.rightRotate(node);
        if (balance < -1 && value >= node.right.value) return this.leftRotate(node);
        if (balance > 1 && value >= node.left.value) {
            node.left = this.leftRotate(node.left);
            return this.rightRotate(node);
        }
        if (balance < -1 && value < node.right.value) {
            node.right = this.rightRotate(node.right);
            return this.leftRotate(node);
        }
        return node;
    }
}
